ext_authz: Set the peer's certificate in the source attributes

sel-vcc opened the pull request:
Signed-off-by: Steve Larkin <steve.larkin@volvocars.com>

Description: Set the downstream client X.509 certificate in the source Peer AttributeContext
Risk Level: low
Testing: Tests updated and extended.
Docs Changes: New API additions are documented.
Release Notes: Added.
Fixes #8326

gsagula commented on source/extensions/filters/common/ext_authz/check_request_utils.cc:
```diff
@@ -68,6 +68,15 @@ void CheckRequestUtils::setAttrContextPeer(envoy::service::auth::v2::AttributeCo
     }
   }
 
+  if (ssl != nullptr && !local) {
```
It looks like that this statement is already being evaluated in the above block: https://github.com/envoyproxy/envoy/blob/e28ecab3a5b9ff304b5b12376d7155d7124ffde7/source/extensions/filters/common/ext_authz/check_request_utils.cc#L43

I think we can avoid unnecessary comparisons by moving this logic into here: https://github.com/envoyproxy/envoy/blob/e28ecab3a5b9ff304b5b12376d7155d7124ffde7/source/extensions/filters/common/ext_authz/check_request_utils.cc#L56

sel-vcc replied:
Thanks! I should have spotted that before uploading!

gsagula commented on source/extensions/filters/common/ext_authz/check_request_utils.cc:
```diff
@@ -68,6 +68,15 @@ void CheckRequestUtils::setAttrContextPeer(envoy::service::auth::v2::AttributeCo
+    const auto peerCertificate = ssl->urlEncodedPemEncodedPeerCertificate();
+    if (!peerCertificate.empty()) {
+      auto& labels = *peer.mutable_labels();
+      labels["peer-certificate"] = peerCertificate;
```
Can you add a test for this, please?

sel-vcc replied:
Done.

sel-vcc commented:
Thanks reviewing, @gsagula!

sel-vcc commented:
@alyssawilk @gsagula @dio Please let me know whether there something further I should do before this can be merged.  Should we add docs and release notes for this change?

sel-vcc commented:
@alyssawilk Is there something further to do with this one?

alyssawilk commented:
Oh sorry, I don't think any maintainer got tagged for review.  @dio can you take a look or should I pick it up Monday?

sel-vcc commented:
@alyssawilk @dio Could you re-open this PR if you are going to take a look?

htuch commented:
@alyssawilk as I understand it, it's not an issue on the security front as we're sending effectively the public key, the client only knows its private key. I think it might be a concern for size though, can we make this a configurable option and turn it off by default? I agree that as we add more attributes, and larger ones, we are going to add to ext authz latency.

alyssawilk commented:
Ok, sounds like fine on the security front, but yeah from priors I'm worried about this carrying everything including the kitchen sink and it being easier to start with an opt-in model rather than not having it configurable, then either having to have 
1) a weird hodgepodge of "legacy things being sent by default" or
2) breaking things where we switch everything to "off by default"
sounds like you're concerned by 2 so @sel-vcc  what do you think about making labels configurable from the get go?

sel-vcc commented:
@alyssawilk Yes, sure, I like that idea.  How would the API look?  It could be as simple as a list of enabled labels, such as:

```diff
diff --git a/api/envoy/config/filter/network/ext_authz/v3alpha/ext_authz.proto b/api/envoy/config/filter/network/ext_authz/v3alpha/ext_authz.proto
index 574fd170d..8f21f030f 100644
--- a/api/envoy/config/filter/network/ext_authz/v3alpha/ext_authz.proto
+++ b/api/envoy/config/filter/network/ext_authz/v3alpha/ext_authz.proto
@@ -31,4 +31,8 @@ message ExtAuthz {
   // communication failure between authorization service and the proxy.
   // Defaults to false.
   bool failure_mode_allow = 3;
+
+  // The labels to provide to the external service in the
+  // :ref:`Peer AttributeContext <envoy_api_msg_service.auth.v3alpha.AttributeContext.Peer`.
+  repeated string enabled_labels = 4;
 }
```
(and the same for the HTTP ext_authz filter)

The well known label names, however, such as `peer-certificate` in this case, are not defined anywhere in the API at present.  I am not sure the best way to define this in protobuf, with string constants not being supported.  Maybe use an enum?  Is there another part of the API from which I can borrow a solution from?

htuch commented:
@sel-vcc I would just add a `repeated string` of labels that are to be sent to the configuration. We don't need to use strict constants.

sel-vcc commented:
Added API and config to make sending the peer labels configurable; squashed previous intermediate commits and rebased against current master.

PTAL @alyssawilk @htuch @gsagula

alyssawilk commented on source/extensions/filters/http/ext_authz/ext_authz.h:
```diff
@@ -74,6 +74,7 @@ class FilterConfig {
         pool_(scope_.symbolTable()),
         metadata_context_namespaces_(config.metadata_context_namespaces().begin(),
                                      config.metadata_context_namespaces().end()),
+        peer_labels_(config.peer_labels().begin(), config.peer_labels().end()),
```
I'm fine with this as is but @htuch any thoughts on the API front on if we want to validate for known values
pros: reduces typos
cons: harder to extend for folks who might subclass ext_authz and add their own custom labels

htuch commented on api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto:
```diff
@@ -90,6 +90,10 @@ message ExtAuthz {
+
+  // The labels to provide to the external service in the
+  // :ref:`AttributeContext.Peer.labels<envoy_api_field_service.auth.v2.AttributeContext.Peer.labels>`.
+  repeated string peer_labels = 10;
```
I think we have insufficient documentation around labels generally, I just had a read through the referenced link and this PR to try and understand this from first principles. Can you explain where these labels are defined, what values are legitimate, etc.?

sel-vcc replied:
I've renamed the field to `labels` and added more documentation, including the legitimate values.

sel-vcc replied:
I have added the docs to the four `ext_authz.proto` files (for http and network filters; v2 and v3alpha versions). That feels correct, because the valid label names could vary per filter-type and API version.  An alternative would have been to document the labels in `attribute_context.proto` but it's not clear to me that this file is specific to external authorization - at least the package name (`envoy.service.auth.v2`) implies that it could be used in other auth cases.

sel-vcc commented on source/extensions/filters/common/ext_authz/check_request_utils.cc:
```diff
@@ -65,6 +73,13 @@ void CheckRequestUtils::setAttrContextPeer(envoy::service::auth::v2::AttributeCo
       } else {
         peer.set_principal(uri_sans[0]);
       }
+      const auto sendPeerCertificate =
```
Oh, I should have spotted that 😊.  Fixed now.

htuch commented on api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto:
```diff
@@ -90,6 +90,16 @@ message ExtAuthz {
+  //
+  // `peer-certificate`: The peer X.509 certificate, in URL- and PEM-encoded format.
+  //
+  // By default no label is sent to the external service.
```
This is much better. Do you think we could do a static doc with this list and avoid repeating here?

I have one other question; if these things are well defined, why don't we just a bunch of bool fields to represent?

sel-vcc replied:
Using bool fields is more explicit.  And if we estimate that there will not be many labels (this is the first) then it might not become too unwieldy to use one field per label.

I've pushed a change using a bool field.  LMK what you think, @htuch.

htuch commented on api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto:
```diff
@@ -90,6 +90,13 @@ message ExtAuthz {
+  // When this field is true, Envoy will include the peer X.509 certificate in the
+  // :ref:`label<envoy_api_field_service.auth.v2.AttributeContext.Peer.labels>` named
+  // `peer-certificate`, in URL- and PEM-encoded format.
+  bool include_peer_certificate = 10;
```
I think this is the right thing where the number of items is bounded and built into the Envoy binary. Why not put it in a dedicated peer certificate field as well? In general, I think static typing is much nicer if we can do it, it seems more comprehensible and easier to program with proto idiomatically vs. a label bag.

sel-vcc replied:
@htuch Something like this?
```diff
diff --git a/api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto b/api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto
index 445c9d7f2..b9359d853 100644
--- a/api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto
+++ b/api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto
@@ -91,12 +91,8 @@ message ExtAuthz {
   // If this field is not specified, the filter will be enabled for all requests.
   api.v2.core.RuntimeFractionalPercent filter_enabled = 9;

-  // Specifies if the peer certificate is sent to the external service.
-  //
-  // When this field is true, Envoy will include the peer X.509 certificate in the
-  // :ref:`label<envoy_api_field_service.auth.v2.AttributeContext.Peer.labels>` named
-  // `peer-certificate`, in URL- and PEM-encoded format.
-  bool include_peer_certificate = 10;
+  // Specifies if and how the peer certificate is sent to the external service.
+  PeerCertificateSettings peer_certificate = 10;
 }

 // Configuration for buffering the request data.
@@ -113,6 +109,14 @@ message BufferSettings {
   bool allow_partial_message = 2;
 }

+// Configuration regarding sending of the peer certificate to the external service.
+message PeerCertificateSettings {
+  // When this field is true, Envoy will include the peer X.509 certificate in the
+  // :ref:`label<envoy_api_field_service.auth.v2.AttributeContext.Peer.labels>` named
+  // `peer-certificate`, in URL- and PEM-encoded format.
+  bool include_with_request = 1;
+}
+
 // HttpService is used for raw HTTP communication between the filter and the authorization service.
 // When configured, the filter will parse the client request and use these attributes to call the
 // authorization server. Depending on the response, the filter may reject or accept the client
```

htuch replied:
More along the lines of replacing the use of labels in the message that gets sent to `ext_authz` (rather than just the config). I.e. we can have a dedicated field for peer certificate in https://github.com/envoyproxy/envoy/blob/a3bc4734e07528abf898ae176e0ba30196d165cd/api/envoy/service/auth/v2/attribute_context.proto#L42

htuch commented:
@sel-vcc can you try and avoid force pushing? This keeps resetting the review history and makes it hard for reviewers to see the incremental diff (also needs master merge..).

dio commented:
LGTM. Modulo fixing the conflicting files. Thanks!

sel-vcc commented:
I have re-resolved the conflict on `version_history.rst`

sel-vcc commented:
/retest

sel-vcc commented:
@htuch @alyssawilk @dio I am not sure what the next step is here: are we set to merge or is further work or approval needed on the API.  (I merged master again, to resolve the version_history conflict, which removed @dio's previous approval.)
